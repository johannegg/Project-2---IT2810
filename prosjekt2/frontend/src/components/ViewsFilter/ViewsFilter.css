.slider {
	display: flex;
	flex-direction: column;
	align-items: center;
	margin-top: 20px;
}

.horizontal-slider {
	width: 100%;
	height: 12px;
	background-color: var(--theme-light-green);
	border-radius: 5px;
}

.thumb {
	height: 12px;
	width: 12px;
	background-color: #493843;
	border-radius: 50%;
	cursor: pointer;
	box-shadow: 0 2px 5px rgba(0, 0, 0, 0.3);
}

.track {
	background-color: var(--theme-dark-green);
}

.viewValues {
	display: flex;
	gap: 15px;
	justify-content: space-between;
	margin-top: 10px;
	color: #493843;
}

@media (max-width: 1050px) {
	.horizontal-slider {
		height: 18px;
	}

	.thumb {
		height: 18px;
		width: 18px;
	}
}

@media (prefers-color-scheme: dark) {
	.viewValues {
		color: white;
	}
}
