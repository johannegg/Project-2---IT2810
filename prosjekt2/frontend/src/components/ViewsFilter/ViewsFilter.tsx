import { useEffect, useState } from "react";
import "./ViewsFilter.css";
import ReactSlider from "react-slider";
import { formatViews } from "../../utils/FormatViews";
import { FaFilter } from "react-icons/fa";

interface ViewsFilterProps {
	onViewsChange: (minViews: number, maxViews: number) => void;
}

export function ViewsFilter({ onViewsChange }: ViewsFilterProps) {
	const [minViews, setMinViews] = useState(0);
	const [maxViews, setMaxViews] = useState(3000000);

	useEffect(() => {
		const savedMinViews = JSON.parse(sessionStorage.getItem("minViews") || "0"); // 0 og 3000000 er hardkodet
		const savedMaxViews = JSON.parse(sessionStorage.getItem("maxViews") || "3000000");
		setMinViews(savedMinViews);
		setMaxViews(savedMaxViews);
	}, []);

	// Updating state while user is dragging the slider
	const handleSliderChange = (values: number[]) => {
		const [newMinViews, newMaxViews] = values;
		setMinViews(newMinViews);
		setMaxViews(newMaxViews);
	};

	// Called when user releases the slider
	const handleAfterSliderChange = (values: number[]) => {
		const [newMinViews, newMaxViews] = values;
		onViewsChange(newMinViews, newMaxViews);
		sessionStorage.setItem("minViews", JSON.stringify(newMinViews));
		sessionStorage.setItem("maxViews", JSON.stringify(newMaxViews));
	};

	return (
		<section className="filterContainer">
			<section className="filterHeader">
				<FaFilter className="filterSortIcon" />
				<h2>Views</h2>
			</section>
			<div className="slider">
				<ReactSlider
					className="horizontal-slider"
					thumbClassName="thumb"
					trackClassName="track"
					min={0}
					max={3000000}
					value={[minViews, maxViews]}
					onChange={handleSliderChange}
					onAfterChange={handleAfterSliderChange}
					renderThumb={(props) => <div {...props} />}
				/>
				<div className="viewValues">
					<span>{formatViews(minViews)} - {formatViews(maxViews)}</span>
				</div>
			</div>
		</section>
	);
}
