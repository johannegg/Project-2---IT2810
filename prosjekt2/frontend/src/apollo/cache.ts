import { InMemoryCache, makeVar } from "@apollo/client";
import { SongData } from "../utils/types/SongTypes";
import { PlaylistData } from "../pages/Playlists/Playlists";

// Initialize sessionStorage and localStorage if values are not already set
const initializeStorage = (key: string, defaultValue: string) => {
  if (!sessionStorage.getItem(key)) {
    sessionStorage.setItem(key, defaultValue);
  }
};

const initializeLocalStorage = (key: string, defaultValue: string) => {
  if (!localStorage.getItem(key)) {
    localStorage.setItem(key, defaultValue);
  }
};

// Initialize values in storage
initializeStorage("minViews", "0");
initializeStorage("maxViews", "1000000");
initializeStorage("selectedGenres", "[]");
initializeStorage("sortOption", "views_desc");
initializeStorage("homeSearchTerm", "");
initializeStorage("favoritesSearchTerm", "");
initializeLocalStorage("favoriteSongs", "[]");
initializeLocalStorage("playlists", "[]");

// Load saved values from storage
const savedGenres = JSON.parse(sessionStorage.getItem("selectedGenres") || "[]");
const savedMinViews = Number(sessionStorage.getItem("minViews"));
const savedMaxViews = Number(sessionStorage.getItem("maxViews"));
const savedSortOption = sessionStorage.getItem("sortOption") || "views_desc";
const savedHomeSearchTerm = sessionStorage.getItem("homeSearchTerm") || "";
const savedFavoritesSearchTerm = sessionStorage.getItem("favoritesSearchTerm") || "";
const savedFavoriteSongs = JSON.parse(localStorage.getItem("favoriteSongs") || "[]");
const savedPlaylists = JSON.parse(localStorage.getItem("playlists") || "[]");

// Create reactive variables
export const genreFilterVar = makeVar<string[]>(savedGenres);
export const sortOptionVar = makeVar<string>(savedSortOption);
export const minViewsVar = makeVar<number>(savedMinViews);
export const maxViewsVar = makeVar<number>(savedMaxViews);
export const homeSearchTermVar = makeVar<string>(savedHomeSearchTerm);
export const favoritesSearchTermVar = makeVar<string>(savedFavoritesSearchTerm);
export const favoriteSongsVar = makeVar<SongData[]>(savedFavoriteSongs);
export const playlistsVar = makeVar<PlaylistData[]>(savedPlaylists);  
export const songDataVar = makeVar<SongData[]>([]);

// Configure cache with reactive variables
const cache = new InMemoryCache({
  typePolicies: {
    Query: {
      fields: {
        genreFilter: {
          read() {
            return genreFilterVar();
          },
        },
        sortOption: {
          read() {
            return sortOptionVar();
          },
        },
        minViews: {
          read() {
            return minViewsVar();
          },
        },
        maxViews: {
          read() {
            return maxViewsVar();
          },
        },
        homeSearchTerm: {
          read() {
            return homeSearchTermVar();
          },
        },
        favoritesSearchTerm: {
          read() {
            return favoritesSearchTermVar();
          },
        },
        favoriteSongs: {
          read() {
            return favoriteSongsVar();
          },
        },
        playlists: {
          read() {
            return playlistsVar();
          },
        },
        songData: {
          read() {
            return songDataVar();
          },
        },
      },
    },
  },
});

// Sync reactive variables with sessionStorage and localStorage
homeSearchTermVar.onNextChange((newHomeSearchTerm) => {
  sessionStorage.setItem("homeSearchTerm", newHomeSearchTerm);
});

favoritesSearchTermVar.onNextChange((newFavoritesSearchTerm) => {
  sessionStorage.setItem("favoritesSearchTerm", newFavoritesSearchTerm);
});

genreFilterVar.onNextChange((newGenres) => {
  sessionStorage.setItem("selectedGenres", JSON.stringify(newGenres));
});

minViewsVar.onNextChange((newMinViews) => {
  sessionStorage.setItem("minViews", String(newMinViews));
});

maxViewsVar.onNextChange((newMaxViews) => {
  sessionStorage.setItem("maxViews", String(newMaxViews));
});

sortOptionVar.onNextChange((newSortOption) => {
  sessionStorage.setItem("sortOption", newSortOption);
});

favoriteSongsVar.onNextChange((newFavorites) => {
  localStorage.setItem("favoriteSongs", JSON.stringify(newFavorites));
});

playlistsVar.onNextChange((newPlaylists) => {
  localStorage.setItem("playlists", JSON.stringify(newPlaylists));
  window.dispatchEvent(new Event("storage")); 
});

export default cache;
