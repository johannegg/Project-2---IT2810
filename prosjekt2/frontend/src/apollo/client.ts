import { ApolloClient } from "@apollo/client";
import cache from "./cache";

const client = new ApolloClient({
	uri: "http://localhost:3001/", // Use this to run locally, for tests
	//uri: "http://it2810-12.idi.ntnu.no:3001", // Use this to run on virtual machine
	cache: cache,
});

export default client;
