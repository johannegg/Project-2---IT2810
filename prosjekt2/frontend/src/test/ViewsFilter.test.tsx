import { render, act } from "@testing-library/react";
import { ViewsFilter } from "../components/ViewsFilter/ViewsFilter";
import { minViewsVar, maxViewsVar, clearFiltersVar } from "../apollo/cache";
import { vi } from "vitest";
import "@testing-library/jest-dom";

// Mock ResizeObserver to prevent errors during testing
beforeAll(() => {
  global.ResizeObserver = class {
    observe() {}
    unobserve() {}
    disconnect() {}
  };
});

describe("ViewsFilter Component", () => {
  // Mock function to simulate the onViewsChange callback
  const mockOnViewsChange = vi.fn();

  // Reset mocks, sessionStorage, and reactive variables before each test
  beforeEach(() => {
    vi.clearAllMocks(); 
    sessionStorage.clear(); 
    minViewsVar(0); 
    maxViewsVar(1000000); 
    clearFiltersVar(false); 
  });

  // Test that ViewsFilter initializes correctly with sessionStorage values
  test("initializes with sessionStorage values", async () => {
    // Set initial values in sessionStorage
    sessionStorage.setItem("minViews", "200");
    sessionStorage.setItem("maxViews", "800000");

    // Render the ViewsFilter component and wrap it in an act block to handle updates
    await act(async () => {
      render(<ViewsFilter onViewsChange={mockOnViewsChange} />);
    });

    // Assert that reactive variables match sessionStorage values
    expect(minViewsVar()).toBe(200);
    expect(maxViewsVar()).toBe(800000);

    // Assert that the mock callback was called with the correct arguments
    expect(mockOnViewsChange).toHaveBeenCalledWith(200, 800000);
  });

  // Test that ViewsFilter matches the saved snapshot
  test("matches snapshot", async () => {
    // Render the ViewsFilter component
    const { container } = render(<ViewsFilter onViewsChange={mockOnViewsChange} />);

    // Assert that the rendered HTML matches the stored snapshot
    expect(container).toMatchSnapshot();
  });
});
