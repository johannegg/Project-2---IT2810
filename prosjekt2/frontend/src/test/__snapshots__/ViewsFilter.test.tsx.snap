// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ViewsFilter Component > matches snapshot 1`] = `
<div>
  <section
    class="filterContainer"
  >
    <section
      class="filterHeader"
    >
      <svg
        class="filterSortIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 512 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
        />
      </svg>
      <h2>
        Views
      </h2>
    </section>
    <div
      class="slider"
      tabindex="-1"
    >
      <div
        class="horizontal-slider"
        style="position: relative;"
      >
        <div
          class="track track-0"
          style="position: absolute; left: 0px; right: 0px;"
        />
        <div
          class="track track-1"
          style="position: absolute; left: 0px; right: 0px;"
        />
        <div
          class="track track-2"
          style="position: absolute; left: 0px; right: 0px;"
        />
        <div
          aria-disabled="false"
          aria-orientation="horizontal"
          aria-valuemax="1000000"
          aria-valuemin="0"
          aria-valuenow="0"
          class="thumb"
          role="slider"
          style="position: absolute; z-index: 1; left: 0px;"
          tabindex="-1"
        />
        <div
          aria-disabled="false"
          aria-orientation="horizontal"
          aria-valuemax="1000000"
          aria-valuemin="0"
          aria-valuenow="1000000"
          class="thumb"
          role="slider"
          style="position: absolute; z-index: 2; left: 0px;"
          tabindex="-1"
        />
      </div>
      <div
        class="viewValues"
      >
        <span>
          0
           - 
          1M
        </span>
      </div>
    </div>
  </section>
</div>
`;
